<!--This is the html script for the output page of the tool. This is implemented using the bootstrap library for layout purposes. Bootstrap utilizes rows and columns for spacial organization. The visualizations are written using the d3 library package. The contents of the page from top to bottoms are:
1) Header that contains title, protein name, uniprot ID, and Gene
2) Sliders: 
  a) Hydropathy cutoff slider with textbox entry underneath that responds to the slider 
  and the visualizations
  b) Minimum blob size slider that responds to the visualizations
3) Mean Hydropathy plot. Blue line adjusts the cutoff on this plot via the hydropathy cutoff slider.
Default is 0.4. To the right is a tool tip that explains the plot when the user hovers over it
4) Blobulation visualization with the Das & Pappu color schena. Tooltip and color key is provided.
5) Blobulation visualization with NCPR color gradient. Tooltip and color bar is provided 
(color bar is just a jpg image)
6) Order and disorder plots (not related to blobulation plots)

4-6 contains SNPs mapped along the sequence from the ClinVar database. User can hover over the SNP,
the SNP will turn red and if the user chooses to click on the SNP the ClinVar data will be open in a new tab 

-->
<html>

<head>
    <!--The blobulation annotations are generated using the D3 visualization library-->
    <script src="https://d3js.org/d3.v4.js"></script>
    <script type="text/javascript"></script>
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="static/style.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>{{ protein_name }} </title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
	<link rel="shortcut icon" href="#">
</head>


<body>
    <div class="container">
		<!--Header that contains Title, protein name, Uni ID, and Gene. 
			Other information can be placed here too? -->
		<div class="text-center">
			<h1> The Protein Blobulator </h1>
			<h6> Uniprot ID: {{ my_uni_id | safe }}</h6>
		</div>
		
		<!--The div id my_dataviz is the div container for ALL D3 visualization plots, color keys-->
		<div id="my_dataviz" class="viz_track">
			<!--This section contains the div containers for the sliders and the text entry box-->
			<div class="text-left">
				Hydropathy cutoff:
				<label style = "opacity: 0.0; display: none;">
					<span id="cutoff_user"></span>
				</label>
				<input type="number" id="cutoff_user_box" min="0" max="1.0" value={{ my_cut | safe }} step=0.001>
				<input type="range" name="mySlider" id="cutoff_user_slider" min="0" max="1.0" value={{ my_cut | safe }} class="myslider1" step=0.001>
			</div>
			
			<div class="text-left">
				Minimum blob size:
				<label style = "opacity: 0.0; display: none;"> <span id="domain_threshold_user"></span>
				</label>
				<input type="number" id="domain_threshold_user_box" min="1" max={{ domain_threshold_max | safe }} value={{ domain_threshold | safe }}   step=1>
				<input type="range" name="mySlider" id="domain_threshold_user_slider" min="1" max={{ domain_threshold_max | safe }} value={{ domain_threshold | safe }} class="myslider2" step=1>
			</div>

			<div>
				<input type="checkbox" class="checkbox" id="mutatebox" value="True"> <label for="mutatebox">Mutate residue</label>
				<!-- my_seq includes single quotes in the string so we have to do this subscripting here -->
				<select id="snp_id">
				{% for residue in my_seq[1:-1] %}
				<option value="{{ loop.index0 + 1 }}"> {{ loop.index0 + 1 }} ({{ residue }})</option>
				{% endfor %}
				</select>
				to
				<select id="residue_type">
					<option disabled>-- Positive --</option>
					<option selected="selected" value="R">R - Arg</option>
					<option value="H">H - His</option>
					<option value="K">K - Lys</option>
					<option disabled>-- Negative --</option> 
					<option value="D">D - Asp</option>
					<option value="E">E - Glu</option>
					<option	disabled>-- Polar --</option>
					<option value="N">N - Asn</option>
					<option value="Q">Q - Gln</option>
					<option value="S">S - Ser</option>
					<option value="T">T - Thr</option>
					<option	disabled>-- Hydrophobic --</option>
					<option value="A">A - Ala</option>
					<option value="I">I - Ile</option>
					<option value="L">L - Leu</option>
					<option value="M">M - Met</option>
					<option value="F">F - Phe</option>
					<option value="W">W - Trp</option>
					<option value="Y">Y - Tyr</option>
					<option value="V">V - Val</option>  
					<option	disabled>-- Other --</option>
					<option value="C">C - Cys</option>
					<option value="G">G - Gly</option>
					<option value="P">P - Pro</option>
				</select>
			</div>

			<div class="text-left">
				<label style = "opacity: 0.0; display: none; position: relative;"> </label>
				<button type="button" id="download_plot_link">Download plots!</button>
				<button type="button" id="download_data_link">Download data!</button>
			</div>
		</div>
	</div>
</body>

<style>
/*CSS styling for the blue i tooltip */
.header {
  line-GLOBAL_HEIGHT: 2rem;
  outline: 1px solid red;
  background: #fff;
  padding: 1rem;

  position: sticky;
  top: 0;
}

.tooltip {
    position: absolute;
    text-align: center;
    width: 100px;
    height: 20px;
    padding: 2px;
    font: 12px sans-serif;
    background: lightsteelblue;
    border: 0px;
    cursor: pointer;
}

.rangeslider {
    width: 90%;
    padding: 10px;
}

/*CSS styling for the top slider*/
.myslider1 {
    -webkit-appearance: none;
    /* Override default CSS styles */
    appearance: none;
    width: 30%;
    /* Full-width */
    height: 20px;
    /* Specified height */
    background: #d3d3d3;
    /* Grey background */
    outline: none;
    /* Remove outline */
    opacity: 0.7;
    /* Set transparency (for mouse-over effects on hover) */
    -webkit-transition: .2s;
    /* 0.2 seconds transition on hover */
    transition: opacity .2s;

}

/*CSS styling for the top slider button*/
.myslider1::-webkit-slider-thumb {
    -webkit-appearance: none;
    cursor: pointer;
    background: #007BFF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
}

.myslider1:hover {
    opacity: 2.0;
}

/*CSS styling for the bottom slider*/
.myslider2 {
    -webkit-appearance: none;
    /* Override default CSS styles */
    appearance: none;
    width: 30%;
    /* Full-width */
    height: 20px;
    /* Specified height */
    background: #d3d3d3;
    /* Grey background */
    outline: none;
    /* Remove outline */
    opacity: 0.7;
    /* Set transparency (for mouse-over effects on hover) */
    -webkit-transition: .2s;
    /* 0.2 seconds transition on hover */
    transition: opacity .2s;

}


/*CSS styling for the bottom slider button*/
.myslider2::-webkit-slider-thumb {
    -webkit-appearance: none;
    cursor: pointer;
    background: #007BFF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
}

.myslider2:hover {
    opacity: 2.0;
}
</style>


<script src="static/figure.js"></script>

<script>

//--------------Javascript for plotting in d3-------------

//VARIABLES
	//tooltips
		var hydropathy_tool_text = ("<p>The linear hydropathy uses a sliding, overlapping window approach to calculate the mean hydropathy for groups of 3 residues. This value is rescaled to lie between 0 (hydrophilic) and 1 (hydrophobic). This plot can help you identify how hydrophobicity varies along your sequence.</p>" +
									"<p>Any stretch of 'minimum blob size' or more residues with mean hydropathy > 'hydropathy cutoff' is classified as a hydrophobic or “h” blob and any remaining stretch of four or more residues is classified as a non-hydrophobic linker or “p” blob.</p>")
		var Das_Pappu_text = "<p>The Das-Pappu phase diagram provides a means to estimate how a disordered sequence might behave based on the charge content. Each blob is colored according to the region they fall in Das-Pappu phase diagram. </p> <img src=\"/static/daspappu_graph.png\" width=\"350\" height=\"350\"> "+
			"<p> <i> <b> Figure adapted from: </b> <a href=\"https://www.pnas.org/content/early/2013/07/29/1304749110\" target=\"_blank\"> Das RK, Papu RV (2013) Conformations of intrinsically disordered proteins are influenced by linear sequence distributions of oppositely charged residues PNAS 110(33):13392–13397.</i> </p>"
		var ncpr_tool_text = "<p>For each blob its net charge per residue (NCPR) is calculated. Each blob is evaluated based on it's fraction of both positively and negatively charged residues. The darker blue a blob is shown here, the higher the fraction of positively charged residues are present within the blob. Alternatively, the darker red a blob is shown here, the higher the fraction of negatively charged residues are present within the blob. An even fraction of positive or negative, or a low fraction of any charged residues, results in a grey color.</p>"
		var uversky_tool_text = "<p>For each blob its distance from the line representing the boundary between ordered and disordered proteins on the Uversky diagram (shown below) is plotted. Postive values are disordered and negative values are ordered. </p> <img src=\"/static/uversky_plot.png\" width=\"375\" height=\"275\"> <p> <i> <b> Figure adapted from: </b> <a href=\"https://pubmed.ncbi.nlm.nih.gov/11025552/\" target=\"_blank\"> Uversky VN, Gillespie JR, Fink AL (2002) Why are \"natively unfolded\" proteins unstructured under physiologic conditions? </i> </p>"

		var enrichment_tool_text = "<p>Here, blobs are colored based upon their enrichment in disease associated SNPs (dSNPs). This idea was investigated in the context of aggregating and non-aggregating proteins at various blob lengths and hydrophobicity cutoffs in a forthcoming paper, from which the figure below is presented (Lohia, et al). </p> <img src=\"/static/dsnp_enrich.png\" width=\"300\" height=\"350\">"
		var disorder_tool_text = "<p>For each blob its fraction of disordered residue is calculated. This disorder calculation is only available if the user uses the Uniprot ID, and uses information provided by the Database of Disordered Protein Prediction. <i> <a href=\"https://academic.oup.com/nar/article/41/D1/D508/1069637\" target=\"_blank\"> Matt E. Oates, Pedro Romero, Takashi Ishida, Mohamed Ghalwash, Marcin J. Mizianty, Bin Xue, Zsuzsanna Dosztányi, Vladimir N. Uversky, Zoran Obradovic, Lukasz Kurgan, A. Keith Dunker, Julian Gough, D2P2: database of disordered protein predictions, Nucleic Acids Research, Volume 41, Issue D1, 1 January 2013, Pages D508–D516, https://doi.org/10.1093/nar/gks1226</i></p>"

	//These variables set fixed values for height and width to be used in making all four visualizations 
		const MARGIN = { top: 30, right: 230, bottom: 30, left: 50 }
		const GLOBAL_WIDTH = 1200 - MARGIN.left - MARGIN.right
		const GLOBAL_HEIGHT = 200 - MARGIN.top - MARGIN.bottom

	//tooltip for SNPs link
		var Tooltip_snps = d3.select("body")
			.append("div") // declare the tooltip div 
			.attr("class", "tooltip")
			.style("opacity", 0);

	//updates the slider/box value entered by the user
		var cutoff_slider_value = document.getElementById("cutoff_user");
		var cutoff_slider = document.getElementById("cutoff_user_slider");

	//Cutoff Box
		var cutoff_box = document.getElementById("cutoff_user_box");

	//Threshold Slider
		var domain_threshold_value = document.getElementById("domain_threshold_user");
		var domain_threshold_slider = document.getElementById("domain_threshold_user_slider");
		
	//Threshold Box
		var domain_threshold_box = document.getElementById("domain_threshold_user_box");
	

	//reads the input variable from flask
		var uni_id = {{ my_uni_id | safe }}
		var uni_id_str = '["' + uni_id + '"]'
		var my_seq = {{ my_seq | safe }}
		var my_data = {{ data.chart_data | safe }}

		var my_cut = {{ my_cut | safe }}
		var my_snp = {{ my_snp | safe }}
		var my_disorder = {{ my_disorder | tojson| safe }}
		var domain_threshold_max = {{ domain_threshold_max | safe }}


	//GLOBAL EVENT HANDLERS
		domain_threshold_value.innerHTML = domain_threshold_slider.value;
		cutoff_slider_value.innerHTML = cutoff_slider.value;
		cutoff_slider.oninput = function() {
			cutoff_box.value = this.value;
			cutoff_slider_value.innerHTML = this.value;
			}
		domain_threshold_box.oninput = function() {
			domain_threshold_slider.value = this.value;
			domain_threshold_value.innerHTML = this.value;
			}
		domain_threshold_slider.oninput = function() {
			domain_threshold_box.value = this.value;
			domain_threshold_value.innerHTML = this.value;
			}
		cutoff_box.oninput = function() {
			cutoff_slider.value = this.value;
			cutoff_slider_value.innerHTML = this.value;
			}
	
	//make the plots
		draw(my_data)

//FUNCTION DECLARATIONS

// A function that updates the charts
function update_Charts(figs, data) {
	figs.pathyFig.update_cutoff_line()

	//iterate over all figures in figs and update them
	for (let [key, value] of Object.entries(figs)){ 
		try{
			value.update_bars(data)
		}
		catch{console.log("Failed to update: "+key)}
	}
}

// Update the output when parameters are changed
function do_update(figs, data) {
	// If the box is check, I mutate the amino acid
	if(d3.select("#mutatebox").property("checked")){
		var snp_id = document.getElementById("snp_id").value;
		var my_seq = {{ my_seq | safe }};
		var amino_acid = document.getElementById("residue_type").value;
		var my_seq =  my_seq.slice(0, snp_id-1) + amino_acid + my_seq.substr(snp_id, my_seq.length);

	// Otherwise I revert back to wild sequence initially uploaded
	} else {
		var my_seq = {{ my_seq | safe }};
	}

	d_value = d3.select("#domain_threshold_user").text()
	selectedValue = d3.select("#cutoff_user").text()
	var m_str = {my_seq: my_seq, domain_threshold: d_value, cutoff: selectedValue, my_disorder:my_disorder}

	$.post( "/postmethod", m_str , "json")
	.then(function(data) {
		var data_tmp = data.chart_data;
		var tt = JSON.stringify(eval("(" + data_tmp + ")"));
		var data_obj = JSON.parse(tt);
		update_Charts(figs, data_obj);
		});
}
    


function draw(data) {
    data.forEach(function(d) {
            d.resid = +d.resid;
            d.hydropathy = +d.hydropathy;
            d.N = +d.N;
            d.hydropathy_3_window_mean = +d.hydropathy_3_window_mean;
            d.domain_to_numbers = +d.domain_to_numbers
            d.m_cutoff = +d.m_cutoff
            d.domain_threshold = +d.domain_threshold
    });
	
	figs = []
	
	figs.pathyFig = new Figure("pathyPlot", data)
		.add_title("Smoothed hydropathy per residue")
		.add_tooltip(hydropathy_tool_text)
		.add_pathy_ylabel()
		.add_cutoff_line()
		.build_barChart()
		.add_xAxis()
		.add_yAxis()
	
	figs.globFig = new Figure("globPlot", data)
		.add_title("Blobs colored according to globular tendency")
		.add_tooltip(Das_Pappu_text)
		.add_GlobularLegend()
		.build_barChart()
		.add_snps()
		.add_phSNP_ylabel()
		.add_xAxis()

	
	figs.ncprFig = new Figure("ncprPlot", data)
		.add_title("Blobs colored according to net charge per residue")
		.add_tooltip(ncpr_tool_text)
		.add_ContinuousLegend("RWB")
		.build_barChart()
		.add_snps()
		.add_phSNP_ylabel()
		.add_xAxis()

	figs.uverskyFig = new Figure("uverskyPlot", data)
		.add_title("Blobs colored according to Uversky diagram")
		.add_tooltip(uversky_tool_text)
		.add_ContinuousLegend("PuOr")
		.build_barChart()
		.add_snps()
		.add_phSNP_ylabel()
		.add_xAxis()

	figs.richFig = new Figure("richPlot", data)
		.add_title("Blobs colored according to dSNP enrichment")
		.add_tooltip(enrichment_tool_text)
		.add_ContinuousLegend("RWB", {min: "0.5", med: "1.0", max: "2.0"})
		.build_barChart()
		.add_snps()
		.add_phSNP_ylabel()
		.add_xAxis()

	if (my_disorder == "0") {
        figs.disorderFig = new Figure("disorderPlot", data)
			.add_title("DISORDER UNAVAILABLE")
			.add_tooltip("Disorder information is acquired from Uniprot. To see the disorder plot, please input the Uniprot ID on the previous page.")
    } else {
    	figs.disorderFig = new Figure("disorderPlot", data)
			.add_title("Blobs colored according to fraction of disordered residue")
			.add_tooltip(disorder_tool_text)
			.add_ContinuousLegend("PuOr")
			.build_barChart()
			.add_snps()
			.add_phSNP_ylabel()
			.add_xAxis()

    }


    var str = "Download plots!";
    var str_j = "Download data!";


    // Set up UI element change handler
	d3.selectAll("#mutatebox,#snp_id,#residue_type,#domain_threshold_user_slider,#cutoff_user_box,#cutoff_user_slider,.checkbox").on("change", function(){do_update(figs, data)});
    

    //download_data
    d3.select('#download_data_link').on('click', function(){
        // If the box is check, I mutate the amino acid
        // TODO: This code should not be duplicated
        var my_seq = {{ my_seq | safe }};
        if(d3.select("#mutatebox").property("checked")){
            var snp_id = document.getElementById("snp_id").value;
            var amino_acid = document.getElementById("residue_type").value;
            my_seq =  my_seq.slice(0, snp_id-1) + amino_acid + my_seq.substr(snp_id, my_seq.length);
       		}

        $.ajax({
            url: "/json",
            type: "post",
            data: {my_seq: my_seq, domain_threshold: d3.select("#domain_threshold_user").text(), cutoff: d3.select("#cutoff_user").text(), my_disorder:my_disorder},
            xhrFields: {responseType: 'blob'},
            success: function(data) {console.log(data)
                var a = document.createElement('a');
                var url = window.URL.createObjectURL(data);
                let name = {{ my_uni_id | safe }};
                a.href = url;
                a.download = name + '_blobulated.csv';
                document.body.append(a);
                a.click();
                a.remove();
                window.URL.revokeObjectURL(url);
            }
        });
    })  

    //download_plots
    $('#download_plot_link').click(function() {
        window.location='/plot?' + 'domain_threshold=' + d3.select("#domain_threshold_user").text() + '&cutoff=' + d3.select("#cutoff_user").text()    
    })
		
};

</script>
<br>
<br>
<br>
<br>
<!--footer for the page-->
<footer>
    <div class="container">

        <p>Protein Blobulator - &copy; Brannigan Lab 2021</p>
    </div>
</footer>

</html>
​
