<!--This is the html script for the output page of the tool. This is implemented using the bootstrap library for layout purposes. Bootstrap utilizes rows and columns for spacial organization. The visualizations are written using the d3 library package. The contents of the page from top to bottoms are:
1) Header that contains title, protein name, uniprot ID, and Gene
2) Sliders: 
  a) Hydropathy cutoff slider with textbox entry underneath that responds to the slider 
  and the visualizations
  b) Minimum blob size slider that responds to the visualizations
3) Mean Hydropathy plot. Blue line adjusts the cutoff on this plot via the hydropathy cutoff slider.
Default is 0.4. To the right is a tool tip that explains the plot when the user hovers over it
4) Blobulation visualization with the Das & Pappu color schena. Tooltip and color key is provided.
5) Blobulation visualization with NCPR color gradient. Tooltip and color bar is provided 
(color bar is just a jpg image)
6) Order and disorder plots (not related to blobulation plots)

4-6 contains SNPs mapped along the sequence from the ClinVar database. User can hover over the SNP,
the SNP will turn red and if the user chooses to click on the SNP the ClinVar data will be open in a new tab 

-->
<html>

<head>
    <!--The blobulation annotations are generated using the D3 visualization library-->
    <script src="https://d3js.org/d3.v4.js"></script>
    <script type="text/javascript"></script>
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="static/style.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>{{ protein_name }} </title>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
	<link rel="shortcut icon" href="#">
</head>


<body>
    <div class="container">
		<!--Header that contains Title, protein name, Uni ID, and Gene. 
			Other information can be placed here too? -->
		<div class="text-center">
			<h1> The Protein Blobulator </h1>
			<h6> Uniprot ID: {{ my_uni_id | safe }}</h6>
		</div>
		
		<!--The div id my_dataviz is the div container for ALL D3 visualization plots, color keys-->
		<div id="my_dataviz" class="viz_track">
			<!--This section contains the div containers for the sliders and the text entry box-->
			<div class="text-left">
				Hydropathy cutoff:
				<label style = "opacity: 0.0; display: none;">
					<span id="cutoff_user"></span>
				</label>
				<input type="number" id="cutoff_user_box" min="0" max="1.0" value={{ my_cut | safe }} step=0.001>
				<input type="range" name="mySlider" id="cutoff_user_slider" min="0" max="1.0" value={{ my_cut | safe }} class="myslider1" step=0.001>
			</div>
			
			<div class="text-left">
				Minimum blob size:
				<label style = "opacity: 0.0; display: none;"> <span id="domain_threshold_user"></span>
				</label>
				<input type="number" id="domain_threshold_user_box" min="1" max={{ domain_threshold_max | safe }} value={{ domain_threshold | safe }}   step=1>
				<input type="range" name="mySlider" id="domain_threshold_user_slider" min="1" max={{ domain_threshold_max | safe }} value={{ domain_threshold | safe }} class="myslider2" step=1>
			</div>

			<div>
				<input type="checkbox" class="checkbox" id="mutatebox" value="True"> <label for="mutatebox">Mutate residue</label>
				<!-- my_seq includes single quotes in the string so we have to do this subscripting here -->
				<select id="snp_id">
				{% for residue in my_seq[1:-1] %}
				<option value="{{ loop.index0 + 1 }}"> {{ loop.index0 + 1 }} ({{ residue }})</option>
				{% endfor %}
				</select>
				to
				<select id="residue_type">
				 <option value="A">A</option>
				 <option value="R">R</option>
				 <option value="N">N</option>
				 <option value="D">D</option>
				 <option value="C">C</option>
				 <option value="Q">Q</option>
				 <option value="E">E</option>
				 <option value="G">G</option>
				 <option value="H">H</option>
				 <option value="I">I</option>
				 <option value="L">L</option>
				 <option value="K">K</option>
				 <option value="M">M</option>
				 <option value="F">F</option>
				 <option value="P">P</option>
				 <option value="S">S</option>
				 <option value="T">T</option>
				 <option value="W">W</option>
				 <option value="Y">Y</option>
				 <option value="V">V</option>                
				</select>
			</div>

			<div class="text-left">
				<label style = "opacity: 0.0; display: none; position: relative;"> </label>
				<button type="button" id="download_plot_link">Download plots!</button>
				<button type="button" id="download_data_link">Download data!</button>
			</div>
		</div>
	</div>
</body>

<style>
/*CSS styling for the blue i tooltip */
.header {
  line-height: 2rem;
  outline: 1px solid red;
  background: #fff;
  padding: 1rem;

  position: sticky;
  top: 0;
}

.tooltip {
    position: fixed;
    text-align: center;
    width: 100px;
    height: 20px;
    padding: 2px;
    font: 12px sans-serif;
    background: lightsteelblue;
    border: 0px;
}

.rangeslider {
    width: 90%;
    padding: 10px;
}

/*CSS styling for the top slider*/
.myslider1 {
    -webkit-appearance: none;
    /* Override default CSS styles */
    appearance: none;
    width: 30%;
    /* Full-width */
    height: 20px;
    /* Specified height */
    background: #d3d3d3;
    /* Grey background */
    outline: none;
    /* Remove outline */
    opacity: 0.7;
    /* Set transparency (for mouse-over effects on hover) */
    -webkit-transition: .2s;
    /* 0.2 seconds transition on hover */
    transition: opacity .2s;

}

/*CSS styling for the top slider button*/
.myslider1::-webkit-slider-thumb {
    -webkit-appearance: none;
    cursor: pointer;
    background: #007BFF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
}

.myslider1:hover {
    opacity: 2.0;
}

/*CSS styling for the bottom slider*/
.myslider2 {
    -webkit-appearance: none;
    /* Override default CSS styles */
    appearance: none;
    width: 30%;
    /* Full-width */
    height: 20px;
    /* Specified height */
    background: #d3d3d3;
    /* Grey background */
    outline: none;
    /* Remove outline */
    opacity: 0.7;
    /* Set transparency (for mouse-over effects on hover) */
    -webkit-transition: .2s;
    /* 0.2 seconds transition on hover */
    transition: opacity .2s;

}


/*CSS styling for the bottom slider button*/
.myslider2::-webkit-slider-thumb {
    -webkit-appearance: none;
    cursor: pointer;
    background: #007BFF;
    border-radius: 20px;
    width: 20px;
    height: 20px;
}

.myslider2:hover {
    opacity: 2.0;
}
</style>
<script>

//--------------Javascript for plotting in d3-------------

//VARIABLES
	//These variables set fixed values for height and width to be used in making all four visualizations 
	var margin = { top: 30, right: 230, bottom: 30, left: 50 },
		width = 1200 - margin.left - margin.right,
		height = 200 - margin.top - margin.bottom;

	//tooltip for SNPs link
	var Tooltip_snps = d3.select("body")
		.append("div") // declare the tooltip div 
		.attr("class", "tooltip")
		.style("opacity", 0);

	//updates the slider/box value entered by the user
	var cutoff_slider_value = document.getElementById("cutoff_user");
	var cutoff_slider = document.getElementById("cutoff_user_slider");

	//Cutoff Box
	var cutoff_box = document.getElementById("cutoff_user_box");

	//Threshold Slider
	var domain_threshold_value = document.getElementById("domain_threshold_user");
	var domain_threshold_slider = document.getElementById("domain_threshold_user_slider");
	
	//Threshold Box
	var domain_threshold_box = document.getElementById("domain_threshold_user_box");
	



//reads the input variable from flask
	var uni_id = {{ my_uni_id | safe }}
	var uni_id_str = '["' + uni_id + '"]'
	var my_seq = {{ my_seq | safe }}
	var my_data = {{ data.chart_data | safe }}
	//console.log(my_seq)


	var my_cut = {{ my_cut | safe }}
	var my_snp = {{ my_snp | safe }}
	var my_disorder = {{ my_disorder | tojson| safe }}
	var domain_threshold_max = {{ domain_threshold_max | safe }}


//GLOBAL EVENT HANDLERS
	domain_threshold_value.innerHTML = domain_threshold_slider.value;
	cutoff_slider_value.innerHTML = cutoff_slider.value;
	cutoff_slider.oninput = function() {
		cutoff_box.value = this.value;
		cutoff_slider_value.innerHTML = this.value;
	}
	domain_threshold_box.oninput = function() {
		domain_threshold_slider.value = this.value;
		domain_threshold_value.innerHTML = this.value;
	}
	domain_threshold_slider.oninput = function() {
		domain_threshold_box.value = this.value;
		domain_threshold_value.innerHTML = this.value;
	}
		cutoff_box.oninput = function() {
		cutoff_slider.value = this.value;
		cutoff_slider_value.innerHTML = this.value;
	}
	
	

	//makes the plots
	draw(my_data)

class Figure {
	constructor(figID) {
	
	}
	
}

//FUNCTION DECLARATIONS
/* add_colorbar
    Function to create linear color bars for graphs. Adapted from: https://www.visualcinnamon.com/2016/05/smooth-color-legend-d3-svg-gradient/
    Inputs: 
      svg container, 
      width and height of the colorbar in px
      min and max values
      x and y position of the upper left corner
      ctop and cend for the first and last colors of the gradient
      Unique id string
      optional arguments: cq1, cmid, cq3 which define the colors at 25%, 50%, and 75% respectively.d
    Returns: 
        none - svg is passed "by reference"
*/
function add_colorbar({svg, width, height, min, med, max, xpos, ypos, ctop, cq1, cmid, cq3, cend, id}={}) {
    svg = add_linearGradient({svg, ctop, cq1, cmid, cq3, cend, id: id})

    //Draw the rectangle and fill with the gradient
    svg.append("rect")
        .attr("width", width)
        .attr("height", height)
        .attr("x", xpos+width) //offset by one width
        .attr("y", ypos-height) //align tops
        .style("fill", "url(#"+id+")");


    //Add labels
    svg.append("text")
        .attr("x", xpos+width+width+6)
        .attr("y", ypos-height+6)
        .text(max)

    if (med){
        svg.append("text")
            .attr("x", xpos+width+width+6)
            .attr("y", (2*ypos-height+12)/2)
            .text(med)
    }

    svg.append("text")
        .attr("x", xpos+width+width+6)
        .attr("y", ypos+6)
        .text(min)
}

/* add_linearGradient:
        Function to create linear color bars for graphs. Adapted from: https://www.visualcinnamon.com/2016/05/smooth-color-legend-d3-svg-gradient/
    Inputs: 
        svg container, 
        ctop and cend for the first and last colors of the gradient
        an id string
        optional arguments: cq1, cmid, cq3 which define the colors at 25%, 50%, and 75% respectively.d
    Returns:
        The svg with a linear gradient with the given id
*/
function add_linearGradient({svg, ctop, cq1, cmid, cq3, cend, id}={}) {
    var defs = svg.append("defs");

    //Append a linearGradient element to the defs and give it a unique id
    var linearGradient = defs.append("linearGradient")
        .attr("id", id)
        .attr("x1", "0%")
        .attr("y1", "0%")
        .attr("x2", "0%")
        .attr("y2", "100%");

    //Define the gradient
        //Set the color for the start (0%)
    linearGradient.append("stop")
        .attr("offset", "0%")
        .attr("stop-color", ctop)

    if (cq1){
        //Set the color for the end (25%)
        linearGradient.append("stop")
            .attr("offset", "25%")
            .attr("stop-color", cq1)
        }

    if (cmid){
        //Set the color for the end (50%)
        linearGradient.append("stop")
            .attr("offset", "50%")
            .attr("stop-color", cmid)
        }

    if (cq3){
        //Set the color for the end (75%)
        linearGradient.append("stop")
            .attr("offset", "75%")
            .attr("stop-color", cq3)
        }
    
        //Set the color for the end (100%)
    linearGradient.append("stop")
        .attr("offset", "100%")
        .attr("stop-color", cend)

    return svg
}

/* add_tooltip
FUNCTION: add_tooltip
SHORT DESCRIPTION: add a small i that provides information to the user in a tooltip
INPUTS:
	svg - a container for a graph (modified by the function directly)
RETURNS:
	none
*/
function add_tooltip(svg, content, {xpos, ypos}={xpos: width, ypos: margin.top}) {
	var xoffset = 300
	infoIcon = svg.append("text")
		.attr("x", xpos)
		.attr("y", ypos)
		.text("i")
		.style("font-size", "20px")
		.style("font-weight", "bold")
		.style("font-family", "zapfino")
		.style("fill", "1E59FC")
		.attr("alignment-baseline", "middle")
		.attr("class", "info-hover")
		
	var Tooltip = d3.select("body")
		.append("div")
		.style("opacity", 0)
		.attr("class", "tooltip")
		.style("background-color", "white")
		.style("border", "solid")
		.style("border-width", "2px")
		.style("border-radius", "5px")
		.style("padding", "5px")
		.style("padding", "5px")
		.style("width", "400px")
		.style("height", "590px")
		.style("font-size", "18px")
		.style("text-align", "justify")
		
	
	infoIcon.on("mouseover", function(d) {
			//console.log(d3.mouse(document.getElementById("my_dataviz"))[1])
			//.style("fill", "black");
			Tooltip.transition()
				.duration(100)
				.style("opacity", 1);
			Tooltip.html(content)
				.style("right", xoffset+"px")
				.style("bottom", 0+"px");
		})
	
	infoIcon.on("mouseout", function(d, i) {
			d3.select(this)
			Tooltip.transition()
				.duration(500)
				.style("opacity", 0);
		})
}

/* add_phSNP_ylabel
FUNCTION: add_phSNP_ylabel
SHORT DESCRIPTION: add an s h SNP text y label to the svg object
INPUTS:
	svg - a container containing an graph modified by the function itself
RETURNS:
	none
*/
function add_phSNP_ylabel(svg) {
	//"p" blob y-axis label for globular tendency plot
	svg.append("text")
		.attr("class", "y label")
		.attr("text-anchor", "middle")
		.attr("x", 0 - (height / 2) - 50)
		.attr("y", margin.left - 80)
		.attr("transform", "rotate(-90)")
		.text("p");

	//"h" blob y-axis label for globular tendency plot
	svg.append("text")
		.attr("class", "y label")
		.attr("text-anchor", "middle")
		.attr("x", 0 - (height / 2))
		.attr("y", margin.left - 80)
		.attr("transform", "rotate(-90)")
		.text("h");

	//"SNPs" y-axis label for globular tendency plot
	svg.append("text")
		.attr("class", "y label")
		.attr("text-anchor", "middle")
		.attr("x", 0 - (height / 2) + 40)
		.attr("y", margin.left - 80)
		.attr("transform", "rotate(-90)")
		.text("SNPs");
		
	return svg
}

/* add_GlobularLegend
FUNCTION: add_GlobularLegend
SHORT DESCRIPTION: add the discrete legend for the globular tendencies graph
INPUTS:
	svg - a container containing an graph modified by the function itself
RETURNS:
	none
*/
function add_GlobularLegend(svg) {
	//This section contains the key that appears next to the Globular tendency colored plot
	//squares for the key

	keysize=20
	var legend = svg.append("g").attr("id", "legend")

	legend.append("rect")
		.attr("x", width + 20)
		.attr("y", margin.top - 25)
		.attr('width', keysize)
		.attr('height', keysize)
		.style("fill", "#0f0")
	legend.append("rect")
		.attr("x", width + 20)
		.attr("y", margin.top + 5)
		.attr('width', keysize)
		.attr('height', keysize)
		.style("fill", "#FEE882")
	legend.append("rect")
		.attr("x", width + 20)
		.attr("y", margin.top + 35)
		.attr('width', keysize)
		.attr('height', keysize)
		.style("fill", "#BF72D2")
	legend.append("rect")
		.attr("x", width + 20)
		.attr("y", margin.top + 65)
		.attr('width', keysize)
		.attr('height', keysize)
		.style("fill", "#f00")
	legend.append("rect")
		.attr("x", width + 20)
		.attr("y", margin.top + 95)
		.attr('width', keysize)
		.attr('height', keysize)
		.style("fill", "#00f")
			
	//Text that appears to the right of the key    
	legend.append("text")
		.attr("x", width + 50)
		.attr("y", margin.top - 15).text("Globular").style("font-size", "15px")
		.attr("alignment-baseline", "middle")
	legend.append("text")
		.attr("x", width + 50)
		.attr("y", margin.top + 15).text("Janus/Boundary").style("font-size", "15px")
		.attr("alignment-baseline", "middle")
	legend.append("text")
		.attr("x", width + 50)
		.attr("y", margin.top + 45).text("Strong Polyelectrolyte").style("font-size", "15px")
		.attr("alignment-baseline", "middle")
	legend.append("text")
		.attr("x", width + 50)
		.attr("y", margin.top + 75).text("Strong Polyanion (-)").style("font-size", "15px")
		.attr("alignment-baseline", "middle")
	legend.append("text")
		.attr("x", width + 50)
		.attr("y", margin.top + 105).text("Strong Polycation (+)").style("font-size", "15px")
		.attr("alignment-baseline", "middle")

}

/* add_snps
*/
function add_snps(svg, x, y) {
	//console.log("not pathyplot")
	svg.append('g')
		.selectAll("rect")
		//.data(my_disorder.map(function(d) { return +d; }))
		.data(my_snp)
		.enter()
		.append("rect")
		.attr("x", function(d) { return x((d.resid)); })
		.attr("y", function(d) { return y(0.8); })
		.attr("width", x.bandwidth())
		.attr("height", function(d) { return y(0.8); })
		.attr("fill", 'black')
		.on("click", function(d) {
			//window.location.href = d.xrefs[0].url+'_blank'
			window.open(d.xrefs.url, '_blank')
		})
		.on("mouseover", function(d) {
			d3.select(this)
				.attr("fill", "red");
			Tooltip_snps.transition()
				.duration(100)
				.style("opacity", .9);
			Tooltip_snps.html(
					'<a href="' + d.xrefs.url + '"target="_blank">' + d.xrefs.id +
					"</a>")
				.style("left", (d3.event.pageX) + "px")
				.style("top", (d3.event.pageY - 28) + "px");
		})
		.on("mouseout", function(d, i) {
			d3.select(this).attr("fill", function() {
				return "" + 'black' + "";
			})
			Tooltip_snps.transition()
				.duration(500)
				.style("opacity", 0);
		})
}

/* add_cutoff
*/
function add_cutoff(svg, x, y, data) {
	var cut_line = svg.append('g')
		.append("path")
		.attr("class", "mypath")
		.datum(data)
		.attr("fill", "none")
		.attr("stroke", "steelblue")
		.attr("stroke-width", 1.5)
		.attr("d", d3.line()
			.x(function(d) { return x(d.resid) })
			.y(function(d) { return y(my_cut) })
		);
	
	return cut_line
}

/* add_hydropathy_bar
*/
function add_hydropathy_bar(svg, x, y, data) {
	var hydropathy_bar = svg.selectAll("mybar")
		.data(data)
		.enter()
		.append("rect")
		.attr("x", function(d) { return x(d.resid); })
		.attr("y", function(d) { return y(d.hydropathy_3_window_mean); })
		.attr("width", x.bandwidth())
		.attr("height", function(d) { return height - y(d.hydropathy_3_window_mean); })
		.attr("fill", 'grey')
		.on("mouseover", function() {
			d3.select(this)
				.attr("fill", "grey");
		})
		.on("mouseout", function(d, i) {
			d3.select(this).attr("fill", function() {
				return "" + 'grey' + "";
			});
		});
}

/*
FUNCTION: CHART
SHORT DESCRIPTION: make svg containers for each plot
INPUTS:
	chartID
	title
RETURNS:
	svg - a d3 container with one svg graph
*/
function chart(chartID, title) {
    //Sets the dimensions and location of the svg container
    var svg = d3.select("#my_dataviz")
		.append("svg")
        .attr("width", width + margin.left + margin.right)
        .attr("height", height + margin.top + margin.bottom)
        .append("g")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
		.attr("id", chartID)

    //Creates the title
    svg.append("text")
        .attr("x", width / 2)
        .attr("y", margin.top - 25)
        .style("text-anchor", "middle")
        .text(title)
        .attr("font-size", "20px")

    return svg
}


function add_legend(svg){
	switch(svg.attr("id")) {
		case "globPlot":
			// Add Legend for globular plot:
			add_GlobularLegend(svg)
			break;
		case "disorderPlot":
			add_colorbar({svg: svg, width: 20, height: 80, xpos: width, ypos: height,
				id: 'disorder_colorbar',
				min: 0, max: 1,
				cend: '#7f3b08',
				cq3: '#ee9d3c', 
				cmid: '#f6f6f7',
				ctop: '#2d004b'
				})
			break;
		case "richPlot":
			//Color bar key to the right of the enrichment plot.
			add_colorbar({svg: svg, width: 20, height: 80, xpos: width, ypos: height,
				id: 'rich_colorbar',
				min: '0', med: '1', max: '2',
				cend: '#ff0000',
				cmid: '#f4f4ff',
				ctop: '#0000ff'
				})
			break;
		case "uverskyPlot":
			add_colorbar({svg: svg, width: 20, height: 80, xpos: width, ypos: height,
				id: 'uversky_colorbar',
				min: '-0.3', max: '0.3',
				cend: '#7f3b08',
				cq3: '#ee9d3c', 
				cmid: '#f6f6f7',
				ctop: '#2d004b'
				})
			break;
		case "ncprPlot":
			//Color bar key to the right of the NCPR plot. It appends a png image.
			add_colorbar({svg: svg, width: 20, height: 80, xpos: width, ypos: height,
				id: 'NCPR_colorbar',
				min: '-1', med: '0', max: '+1',
				cend: '#ff0000',
				cmid: '#f4f4ff',
				ctop: '#0000ff'
				})
			break;
		default:
			try{throw svg.attr("id")} catch(e) {console.log("I don't know how to make that legend: "+e)}
	}
}

function make_pathy_ylabel(svg) {
	//Creates the "Mean Hydropathy" y-axis label for Smoothed hydropathy per residue
	svg.append("text")
		.attr("class", "y label")
		.attr("text-anchor", "middle")
		.attr("x", 0 - (height / 2))
		.attr("y", margin.left - 80)
		.attr("transform", "rotate(-90)")
		.text("Mean Hydropathy");
}

function build_barChart(svg, data, x, y) {
	if (svg.attr("id") == "pathyPlot") {
		svg.append("g") //the y axis is drawn only for plot 1
			.call(d3.axisLeft(y));
	} else {
		var barChart = svg.append("g")
		barChart.attr("id", "barChart"+svg.attr("id"))
		var plot_variable = barChart.selectAll("rect")
			.data(data)
			
		make_bars(plot_variable, x, y)
	}
	return plot_variable
}



function make_bars(plot_variable, x, y) {
	plot_variable.enter()
		.append("rect")
		.merge(plot_variable)
		.transition()
		.duration(1000)
		.attr("x", function(d) { return x(d.resid); })
		.attr("y", function(d) { return y(d.domain_to_numbers); })
		.attr("width", x.bandwidth())
		.attr("height", function(d) { return height - y(d.domain_to_numbers); })
		.attr("fill", function(d) { return d.var; })
}

function make_cut_line(svg, x, y, data) {
	var cut_line = add_cutoff(svg, x, y, data)
	add_hydropathy_bar(svg, x, y, data)
	return cut_line
}

function add_xAxis(svg, x, y){
	var xAxis = svg.append("g")
	.call(d3.axisBottom(x).tickValues(x.domain().filter(function(d, i) { return !((i+1) % 
	   (Math.round((Math.round(domain_threshold_max/10))/10)*10) )})))
	.attr("transform", "translate(0," + height + ")");
	// Bars
	
	    //Creates the "Residue" x-axis label
    svg.append("text")
        .attr("x", width / 2)
        .attr("y", height + margin.bottom)
        .style("text-anchor", "middle")
        .text("Residue")
	
	return xAxis
}

// A function that update the chart when eitgher slider is moved
function updateCharts(given_cutoff, data_updated, cut_line, x, y, plot_list, plot_variables) {
	cut_line
		.transition()
		.duration(1000)
		.attr("d", d3.line()
			.x(function(d) { return x(d.resid); })
			.y(function(d) { return y(given_cutoff); })
		);

	var name
	for (let [index, svg] of plot_list.entries() ) {
		//console.log(index)
		name = svg.attr("id")
		if  (name=="globPlot") {
			//console.log("update globs")
			data_updated.forEach(function(d) {
			d.var = d.P_diagram
			})
		} else if (name=="ncprPlot") {
			data_updated.forEach(function(d) {
			d.var = d.NCPR_color
			})
		} else if (name=="richPlot") {
			data_updated.forEach(function(d) {
			d.var = d.h_blob_enrichment
			})
		} else if  (name=="disorderPlot") {
			data_updated.forEach(function(d) {
			d.var = d.disorder_color
			})
		} else {
			data_updated.forEach(function(d) {
			d.var = d.uversky_color
			})
		}

		if (svg.attr("id")!="pathyPlot") {
			d3.select("#"+"barChart"+name).selectAll("rect")
				.data(data_updated)
			console.log("nameSlice: "+name.slice(0,-4)+"Bars")
			var plot_variable = plot_variables[name.slice(0,-4)+"Bars"]
				console.log(plot_variables.globBars)
			plot_variable = make_bars(plot_variable, x, y)

		} else {}
	}
}

// Update the output when parameters are changed
function do_update(cut_line, x, y, plot_list, plot_variables) {
	//console.log("do_update")
	// If the box is check, I mutate the amino acid
	if(d3.select("#mutatebox").property("checked")){
		var snp_id = document.getElementById("snp_id").value;
		var my_seq = {{ my_seq | safe }};
		var amino_acid = document.getElementById("residue_type").value;
		var my_seq =  my_seq.slice(0, snp_id-1) + amino_acid + my_seq.substr(snp_id, my_seq.length);

	// Otherwise I revert back to wild sequence initially uploaded
	} else {
		var my_seq = {{ my_seq | safe }};
	}

	var d_value = d3.select("#domain_threshold_user").text()
	var selectedValue = d3.select("#cutoff_user").text()
	var m_str = {my_seq: my_seq, domain_threshold: d_value, cutoff: selectedValue, my_disorder:my_disorder}

	$.post( "/postmethod", m_str , "json")
	.then(function(data) {
		var data_tmp = data.chart_data;
		var tt = JSON.stringify(eval("(" + data_tmp + ")"));
		var data_obj = JSON.parse(tt);
		updateCharts(selectedValue, data_obj, cut_line, x, y, plot_list, plot_variables);
	});
}
    

function draw(data) {
	
	//tooltips
	var hydropathy_tool_text = ("<p>The linear hydropathy uses a sliding, overlapping window approach to calculate the mean hydropathy for groups of 3 residues. This value is rescaled to lie between 0 (hydrophilic) and 1 (hydrophobic). This plot can help you identify how hydrophobicity varies along your sequence.</p>" +
									"<p>Any stretch of 'minimum blob size' or more residues with mean hydropathy > 'hydropathy cutoff' is classified as a hydrophobic or “h” blob and any remaining stretch of four or more residues is classified as a non-hydrophobic linker or “p” blob.</p>")

	var Das_Pappu_text = "<p>The Das-Pappu phase diagram provides a means to estimate how a disordered sequence might behave based on the charge content. Each blob is colored according to the region they fall in Das-Pappu phase diagram. </p> <img src=\"/static/daspappu_graph.png\" width=\"350\" height=\"350\"> <p> <i> <b> Figure adapted from: </b> <a href=\"https://www.pnas.org/content/early/2013/07/29/1304749110\" target=\"_blank\"> Das RK, Papu RV (2013) Conformations of intrinsically disordered proteins are influenced by linear sequence distributions of oppositely charged residues PNAS 110(33):13392–13397.</i> </p>"
	var NCPR_tool_text = "<p>For each blob its net charge per residue (NCPR) is calculated. </p>"
	var uversky_tool_text = "<p>For each blob its distance from the uversky diagram slope is plotted. Postive values are disordered and negative values are ordered. </p>"
	var enrichment_tool_text = "<p>Enrichment values for h blobs</p>"
	var disorder_tool_text = "<p>For each blob its fraction of disorderd residue is calculated.</p>"

//setup and preprocess data
    var domain_bar, domain_bar_disorder,domain_bar_ncpr, domain_bar_u;

    var plot_variables = {"globBars": domain_bar, "ncprBars": domain_bar_ncpr, "disorderBars": domain_bar_disorder, "uverskyBars": domain_bar_u };

    var tem = plot_variables.globBars

    data.forEach(function(d) {
            d.resid = +d.resid;
            d.hydropathy = +d.hydropathy;
            d.N = +d.N;
            d.hydropathy_3_window_mean = +d.hydropathy_3_window_mean;
            d.domain_to_numbers = +d.domain_to_numbers
            d.m_cutoff = +d.m_cutoff
            d.domain_threshold = +d.domain_threshold
    });
	
	var y = d3.scaleLinear()
		.domain([0, 1])
		.range([height, 0]);

	// add the x Axis
	var x = d3.scaleBand()
		.range([0, width])
		.domain(data.map(function(d) { return d.resid; }))
		.padding(0.2);

//Make graphs
	
    var pathySVG = chart("pathyPlot", "Smoothed hydropathy per residue")
		add_xAxis(pathySVG, x, y)
		build_barChart(pathySVG, data, x, y)
		var cut_line = make_cut_line(pathySVG, x, y, data)
		make_pathy_ylabel(pathySVG)
		add_tooltip(pathySVG, hydropathy_tool_text)
	
    var globsSVG = chart("globPlot", "Blobs colored according to globular tendency")
		data.forEach(function(d) {d.var = d.P_diagram})
		add_xAxis(globsSVG, x, y)
		plot_variables.globBars = build_barChart(globsSVG, data, x, y)
		add_snps(globsSVG, x, y)
		add_phSNP_ylabel(globsSVG)
		add_tooltip(globsSVG, Das_Pappu_text)
		add_legend(globsSVG)
		
    var ncprSVG = chart("ncprPlot", "Blobs colored according to net charge per residue")
		data.forEach(function(d) {d.var = d.NCPR_color})
		plot_variables.ncprBars = build_barChart(ncprSVG, data, x, y)
		add_snps(ncprSVG, x, y)
		add_phSNP_ylabel(ncprSVG)
		add_xAxis(ncprSVG, x, y)
		add_tooltip(ncprSVG, NCPR_tool_text)
		add_legend(ncprSVG)
		
    var uverskySVG = chart("uverskyPlot", "Blobs colored according to uversky diagram")
		data.forEach(function(d) {d.var = d.uversky_color})
		plot_variables.uverskyBars = build_barChart(uverskySVG, data, x, y)
		add_snps(uverskySVG, x, y)
		add_phSNP_ylabel(uverskySVG)
		add_xAxis(uverskySVG, x, y)
		add_tooltip(uverskySVG, uversky_tool_text)
		add_legend(uverskySVG)
		
    var richSVG = chart("richPlot", "Blobs colored according to enrichments")
		data.forEach(function(d) {d.var = d.h_blob_enrichment})
		plot_variables.richBars = build_barChart(richSVG, data, x, y)
		add_snps(richSVG, x, y)
		add_phSNP_ylabel(richSVG)
		add_xAxis(richSVG, x, y)
		add_tooltip(richSVG, enrichment_tool_text)
		add_legend(richSVG)
		
	if (my_disorder == "0") {
        var plot_list = [pathySVG, globsSVG, ncprSVG, uverskySVG, richSVG]
    } else {
        var disorderSVG = chart("disorderPlot", "Blobs colored according to fraction of disordered residue")
        data.forEach(function(d) {d.var = d.disorder_color})
		plot_variables.disorderBars = build_barChart(disorderSVG, data, x, y)
		add_phSNP_ylabel(disorderSVG)
		add_snps(disorderSVG, x, y)
		add_xAxis(disorderSVG, x, y)
		add_tooltip(disorderSVG, disorder_tool_text)
		add_legend(disorderSVG)
		var plot_list = [pathySVG, globsSVG, ncprSVG, disorderSVG, uverskySVG, richSVG] 
    }

console.log(plot_variables.globBars)
    var str = "Download plots!";
    var str_j = "Download data!";


    // Set up UI element change handlers
    d3.selectAll(".checkbox").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)})
    d3.select("#cutoff_user_slider").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)});
    d3.select("#cutoff_user_box").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)});
    d3.select("#domain_threshold_user_slider").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)});
    d3.select("#domain_threshold_user_box").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)});
    d3.selectAll("#mutatebox,#snp_id,#residue_type").on("change", function(){do_update(cut_line, x, y, plot_list, plot_variables)});

	
    //download_data
    d3.select('#download_data_link').on('click', function(){
        // If the box is check, I mutate the amino acid
        // TODO: This code should not be duplicated
        var my_seq = {{ my_seq | safe }};
        if(d3.select("#mutatebox").property("checked")){
            var snp_id = document.getElementById("snp_id").value;
            var amino_acid = document.getElementById("residue_type").value;
            my_seq =  my_seq.slice(0, snp_id-1) + amino_acid + my_seq.substr(snp_id, my_seq.length);
        }

        $.ajax({
                url: "/json",
                type: "post",
                data: {my_seq: my_seq, domain_threshold: d3.select("#domain_threshold_user").text(), cutoff: d3.select("#cutoff_user").text(), my_disorder:my_disorder},
                xhrFields: {
                responseType: 'blob'
                },
                success: function(data) {console.log(data)
                var a = document.createElement('a');
                var url = window.URL.createObjectURL(data);
                a.href = url;
                a.download = 'myfile.csv';
                document.body.append(a);
                a.click();
                a.remove();
                window.URL.revokeObjectURL(url);
                }
            });
        })  

    //download_plots
    $('#download_plot_link').click(function() {
        window.location='/plot?' + 'domain_threshold=' + d3.select("#domain_threshold_user").text() + '&cutoff=' + d3.select("#cutoff_user").text()    
        })
};

</script>
<br>
<br>
<br>
<br>
<!--footer for the page-->
<footer>
    <div class="container">

        <p>Protein Blobulator - &copy; Brannigan Lab 2021</p>
    </div>
</footer>

</html>
​