<!--This is the html script for the opening page of the tool. This is implemented using the bootstrap library.
  Bootstrap utilizes rows and columns for spacial organization.
The opening page contains three tabs: Home, About, and Documentation.-->
<!DOCTYPE html>
<html lang="en">
<meta charset="utf-8">

<head>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>Blobulator &copy;</title>
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"></script>
    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
    ​
    ​
    <style>
        .bs-example{
        margin: 20px; }
    
img {
  display: block;
  margin-left: auto;
  margin-right: auto;
}
</style>
</head>

<body>
    <!--This is a row that contains the header with two images of the BDNF illustrations
  on each side of the text-->
    <div class="container">
        <div class="row justify-content-center">
            <div class="col-xs-6-4">
                <img src="/static/IDP_decor1.png" width="120" height="125">
            </div>
            <div class="col-xs-6">
                <h2>The Protein Blobulator</h2>
            </div>
            <div class="col-xs-6">
                <img src="/static/IDP_decor2.png" width="84" height="125">
            </div>
        </div>
    </div>
    <!--This creates a row to display the clickable toggling tabs for Home, About, and Documentation-->
    <div class="bs-example">
        <ul class="nav nav-tabs">
            <li class="nav-item">
                <a href="#home" class="nav-link" data-toggle="tab">Home</a>
            </li>
            <li class="nav-item">
                <a href="#about" class="nav-link" data-toggle="tab">About</a>
            </li>
            <li class="nav-item">
                <a href="#documentation" class="nav-link" data-toggle="tab">Documentation</a>
            </li>
        </ul>
        ​
        <div class="tab-content">
            ​
            <!--About Page Contents-->
            <div class="tab-pane fade" id="about">
                <h4 class="mt-2">Welcome to the Protein Blobulator</h4>
                <p>The Protein Blobulator is an online resource developed at The<a href="https://branniganlab.wordpress.com/news-2/" target="_blank"> Brannigan Lab</a> at Rutgers-Camden Center for Computational and Integrative Biology. The goal of the Protein Blobulator is to annotate and illustrate subdomains of proteins that have distinct physical properties from each other, and make predictions about how these distinctions have functional effects. <br> <br> The information provided on this website spans any protein sequence or Uniprot ID, but it is especially useful for Intrinsically Disordered Proteins. This website also provides information about Disease-associated SNPs that have been obtained from the <a href="https://www.ncbi.nlm.nih.gov/clinvar/" target="_blank">Clin Var</a> database.</p>
                <br><br>
                <p> What is blobulation? The nomenclature used here comes from polymer physics, where the segments of a polypeptide chain are grouped together into interaction domains, where the residues within a certain region of the protein can be expected to behave in a relatively predictable way (de Gennes, 1979).</p>
                <br><br> For bug reports, feature requests, or anything else please contact at grace.brannigan@rutgers. edu.</p>
                <br>Citing Blobulator<br> <p>A manuscript is currently in preparation. For now, please cite analysis done with Blobulator using the abstract from the Biophysical Society meetingas follows: Article (Lohia2020) Lohia, R.; Bassi, K.; Hansen, M. & Brannigan, G. (2020). Disease Associated Mutations in Intrinsically Disordered Proteins Show Evidence of Enrichment in Hydrophobic Blobs. Biophysical Journal, Elsevier BV,118, 215a.</p>

            </div>
            <!--Example Page Contents-->
            <div class="tab-pane fade" id="documentation">
                <h4 class="mt-2">A Quick Overview</h4>
                <p> The first thing needed for the blobulator is either a Uniprot ID for a protein of interest, or a manual sequence entry. Uniprot IDs are our recommended format for retrieving the sequence required for blobulation. Above each “compute button” there is a manual text entry box, in which the user should insert either their sequence or ID. The user should then press the “Compute” button. </p>
                <br> <br>
                <p> On the blobulation output page, there are three adjustable parameters, all of which can be adjusted by the user: </p>
                <p> <b> Adjusting Hydropathy Cutoff </b> </p>
                <p> The first of these is the hydropathy cutoff. Blobulation considers the residues in turn and calculates the average hydrophobicity score of each group of residues. If a group of residues (under the minimum blob size) is under the value for hydrophobicity cutoff, it will be considered a ‘h’ (hydrophobic) blob, and if it’s below it will be considered a ‘p’ (non-hydrophobic) blob. The hydropathy cutoff can be adjusted by manual entry into the text box to the left, or by adjusting the slider to the desired value. </p>
                <p> <b> Adjusting Minimum Blob Size </b> </p>
                <p> The second of the adjustable parameters on the output page is the minimum blob size. This sets a threshold for how many residues constitute blob. If there are groups of residues which average above or below the hydropathy cutoff, they must then meet this requirement to be considered either an h or p blob. If a residue in a h-blob is below this threshold and below the hydropathy cutoff, or if the residue is in a p-blob and above the hydropathy cutoff and below the minimum blob size, the residue (or group of residues) will be considered a ‘s’ (spacer or separator) blob. </p>
                <p> <b> Mutating Residues </b> </p>
                <p> The third and final adjustable parameter on the blobulation output page is the mutate residue option. This allows the user to change a residue within the sequence, and see what potential affects it would have on the blobulation output. To mutate a residue, the user should check the Mutate Residue box on the left, then select which residue they would like to mutate, and then choose the amino acid they would like to mutate the residue into. The graphs will automatically update after the box is checked and the changes are made. </p>
                <br> <br>
                <p>After blobulation, multiple graphs are produced and available to the user. </p>
                <p> <b> Smoothed hydropathy per residue </b> </p>
                <p> The first graph displays the smoothed hydropathy per residue. The core of blobulation consists of two parameters - the first of which being a hydropathy threshold. This threshold is shown by the blue line on the “mean hydropathy” axis. This graph is the only one that shows the residues individually, and can be used as a reference to how the residues are grouped together based upon their position above or below the mean hydropathy line. </p>
                <p> <b> Blobs according to globular tendency </b> </p>
                <p> The second graph shows the blobs according to their globular tendency, considered based upon their Das-Pappu classification. Here the residues are grouped based upon their determined blobs, and SNP data is shown above the relevant residues as black lines.</p>
                <p> <b> Blobs according to net charge per residue </b> </p>
                <p> The third graph shows the blobs according to their residues’ collective average charge. Here the residues are again grouped based upon their determined blobs, and SNP data is shown above the relevant residues as black lines. </p>
                <p> <b> Blobs colored according to Uversky diagram </b> </p>
                <p> The fourth graph shows the blobs according to their positions on the Uversky diagram, where the line between ordered and disordered is plotted. Here the residues are grouped based upon their determined blobs, and SNP data is shown above the relevant residues as black lines. </p>
                <p> <b> Blobs colored according to dSNP enrichment </b> </p>
                <p> The fifth graph shows the blobs according to their enrichment in documented disease associated SNPs (dSNP). Here the residues are grouped based upon their determined blobs, and SNP data is shown above the relevant residues as black lines. </p>
                <p> <b> Blobs colored according to fraction of disordered residue </b> </p>
                <p> The sixth and final graph shows the blobs according to their fraction of disordered residue, which utilizes the Database of Disordered protein prediction. Here the residues are grouped based upon their determined blobs, and SNP data is shown above the relevant residues as black lines. </p>
                <p> <b> Downloading Data </b> </p>
                <p> After the “Download data!” button (located just below the three adjustable parameters) is pressed, the raw data will be downloaded in the form of a csv file. The data stored here is by residue. Each column corresponds to one of the following: residue name, residue number, window, hydropathy cutoff, minimum blob size, average hydrophobicity, blob type, blob index number, blob Das-Pappu classification, blob net charge per residue, fraction of positively charged residues, fraction of negatively charged residues, fraction of charged residues, uversky diagram score, blob dSNP enrichment, and blob disorder score. </p>
                <br> <br>
                <h4 class="mt-2">FAQ:</h4>
                <br> <br>
                <p> <b> <i> What does the blobulator output? </i> </b> </p>
                <p> The blobulator outputs 6 graphs showing categorizations of its subdomains. Each graph shows the sequence of the protein displayed in one of the following ways: smoothed hydropathy per residue, colored according to globular tendency, colored according to net charge per residue, colored according to the Uversky diagram, colored according to dSNP enrichment, colored according to fraction of disordered residue. </p>
                <p> <b> <i> Do you recommend using the manual sequence entry option, or the Uniprot ID option for input? </i> </b> </p>
                <p> We strongly recommend using the Uniprot ID option when available. There will be more graphs outputted, as well as SNP data available if you blobulate the protein using its Uniprot ID. If you are interested in a specific variant of the protein, such as one containing a SNP, there is a mutate residue option at the top of the output page. </p>
                <p> <b> <i> Can I download the plots? </i> </b> </p>
                <p> Yes! We recommend saving the page as a .pdf file using the “print” function in your browser. </p>
                <p> <b> <i> How do I access the data used to generate the plots? </i> </b> </p>
                <p> The data can be downloaded using the “Download data!” button at the top of the blobulator output page. The downloaded data will be in the form of the csv file with labeled columns, which can be used to generate custom graphs or retrieve specific values. </p>
                <p> <b> <i> Will the downloaded data change if I change the hydropathy cutoff or blob side using the sliders? </i> </b> </p>
                <p> Yes. Any adjustments made after blobulation but before the “Download data!” button is clicked will be reflected in the csv file. </p>
                <p> <b> <i> Why am I not seeing SNP data? </i> </b> </p>
                <p> It is possible that you chose the manual sequence entry option, for which there will be no SNP data, or that there is no data in EMBL-EBI for your protein of interest. In either case, it doesn’t necessarily mean that no SNP data exists for the protein you are blobulating. </p>
                <p> <b> <i> Why isn’t the Blobs colored according to fraction of disordered residue graph not showing up on my output page? </i> </b> </p>
                <p> Check and make sure you chose the Uniprot ID input option. </p>
                <p> <b> <i> I would really like to see a certain feature in the next version. </i> </b> </p>
                <p> Please contact us and let us know what you’re thinking. Our goal is to maximize the blobulator’s usefulness, and any suggestions are greatly appreciated. In the meantime, the local version of the blobulator, which can be found on our github, can be modified to your liking. </p>


            </div>
            ​
            ​
            <!--Home Page Contents-->
            <div class="tab-pane container active" id="home">
                <h5 class="text-center">Please enter the <a href="https://www.uniprot.org" target="_blank"> Uniprot ID </a> or amino acid sequence for blobulation.</h5>
                <br>
                <br>
                <div class="container-fluid">
                    <div class="row">
                        <div class="col-sm-4">
                            <!--form-->
                            <div class="form-group">
                                <form class="navbar-form navbar-top" method=post target="_blank" action="">
                                    <label> {{form.uniprot_id.label }}</label>
                                    <p> The UniProtKB accession number is the unique alphanumeric identifier of each entry in UniProtKB. </p>
                                    <!--form.uniprot_id calls the Uniprot ID field in model_2.py -->
                                    <p> {{ form.uniprot_id(class_="form-control", rows="1")}}</p>
                                    <p><input type=submit name="action_u" class="btn btn-primary" value="Compute"></p>
                                    <p>
                                        {% if result != None %}
                                        {% endif %}
                                    </p>
                                </form>
                            </div>
                        </div>
                        <div class="col-sm-2 align-self-center">
                            <div class="text-center">
                                <h2>Or</h2>
                            </div>
                        </div>
                        <div class="col-sm-4">
                            <div class="form-group">
                                <form class="navbar-form navbar-top" method=post target="_blank" action="">
                                    <label> {{form.aa_sequence.label }}</label>
                                    <p> Input any sequence of your choosing. </p>
                                    <p> The user should note that if they choose to enter their sequence manually that certain outputs, such as SNP identification and display of blobs according to fraction of disordered residue, will be omitted. Blobulation using Uniprot ID is recommended. <p>
                                    <p> {{ form.aa_sequence(class_="form-control", rows="5")}}</p>
                                    <label> {{form.seq_name.label }}</label>
                                    <p> Enter any name for your sequence of choice. </p>
                                    <p> {{ form.seq_name(class_="form-control", rows="1")}} </p>
                                    <p><input type=submit name="action_s" class="btn btn-primary" value="Compute"></p>
                                    <p>
                                        {% if result != None %}
                                        {% endif %}
                                    </p>
                            </div>
                        </div>
                        ​
                    </div>
</body>
​
​
​
<!--a function that disables the enter key from adding new lines to the textarea-->
<script type="text/javascript">
$(document).ready(function() {

    $('p').keypress(function(event) {

        if (event.keyCode == 13) {
            event.preventDefault()
        }
    });
});
</script>
​
​
​
<br>
<br>
<br>
​
<!--footer for the page-->
<footer>
    <div class="container">
        <p>Protein Blobulator - &copy; Brannigan Lab 2019</p>
    </div>
</footer>
​

</html>